### ISSUE TITLE
Why clearing all cached messages when session is clean ?

### ISSUE THREAD

**Issue by alaeddineG** [ckrey/MQTT-Client-Framework#185]
Hi,

This is a question.

I just wanted to ask why the session is clearing the persisted message or flow as you name it when the session is clean ?
Is there a way to go around this `[self.persistence deleteAllFlowsForClientId:self.clientId];` ?

    - (void)connect {
        DDLogVerbose(@"[MQTTSession] connecting");
        if (self.cleanSessionFlag) {
            [self.persistence deleteAllFlowsForClientId:self.clientId];
            [self.subscribeHandlers removeAllObjects];
            [self.unsubscribeHandlers removeAllObjects];
            [self.publishHandlers removeAllObjects];
        }
        [self tell];
        
        self.status = MQTTSessionStatusConnecting;
        
        self.decoder = [[MQTTDecoder alloc] init];
        self.decoder.runLoop = self.runLoop;
        self.decoder.runLoopMode = self.runLoopMode;
        self.decoder.delegate = self;
        [self.decoder open];
        
        self.transport.delegate = self;
        [self.transport open];
    }

Because I'am connected to the broker as clean and since I'am using MQTTClient to publish the client location : whenever a significant location change is fired I send the location payload using this :

    - (UInt16)sendData:(NSData *)data topic:(NSString *)topic qos:(MQTTQosLevel)qos retain:(BOOL)retainFlag
    {
        if (self.state != MQTTSessionManagerStateConnected) {
            [self connectToLast];
        }
        UInt16 msgId = [self.session publishData:data
                                         onTopic:topic
                                          retain:retainFlag
                                             qos:qos];
        return msgId;
    }

When the connection is not clean and I send a location payload after significant location change the message is delivered by when the connection is clean the message is not.
So basically if the session is clean the session object will remove all the persisted message when it connect to the broker and the persisted message are deleted.

Is there a way to get advantage of this persistance layer when the clean session is set ? Or I have to cash by myself those locations and send them when the broker is connected ?

Please let me know if my logic (interpretation) is wrong or I misunderstood the the why the persistance work and clear the message.

Thank you :)

**Comment by ckrey** [comment]
An MQTT client can only assure that a message has been successfully delivered by checking the PUBLISH/PUBACK or PUBLISH/PUBREC/PUBCOMP/PUBREL protocol handshake in QoS1 or QoS2 with the broker.

If the broker does not support persistent sessions (cleansession=NO), there is no way to tell if the broker has received a message after a connection problem and reconnect.

Yes, in that case you have to do queueing/acknowledgement on application level like you would do with HTTP.

**Comment by alaeddineG** [comment]
Thank you ckrey for your quick answer.

Here is how I understood it :
So in case the session is clean and the client send a message with Qos 1 or 2 and disconnect before receiving and acknowledgement from the broker there is no way to tell if this message have been delivered or not. That's mean if session is not clean and client send a message with Qos 1 or 2 and disconnect and reconnect he would receive an acknowledgement when it' get connected back (I assume the client should maintain the same client id i guess ).
Please correct me If I'am getting this wrong. :)

Since my connection is clean if I send a message while not connected it will be lost when the client is connecting because it will need to remove all persisted message because there no why to know if those messages have been delivered since the client wont get any acknowledgement back when he connect back to the broker.

In this case if my client is only publishing data what is the advantage of connecting with a clean session ? ( my client is not subscribing to any topic for now )

Thank you again ckrey !!

**Comment by ckrey** [comment]
The QoS1 or Qos2 handshake sequence is the same no matter of the direction of the transfer. This means, to recover from an error while transmitting to the server you need Cleansession=No too, even if you are not subscribing to any topic.

If you comment the code you may be transmitting duplicates
